segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/mocks/InMemoryChunkStorage.java

```diff
@@ -0,0 +1,295 @@
+    }
+
+    @Override
+    protected ChunkHandle doOpenRead(String chunkName) throws ChunkStorageException, IllegalArgumentException {
+        if (chunks.containsKey(chunkName)) {
+            return new ChunkHandle(chunkName, true);
+        }
+        throw new ChunkNotFoundException(chunkName, "InMemoryChunkStorage::doOpenRead");
+    }
+
+    @Override
+    protected ChunkHandle doOpenWrite(String chunkName) throws ChunkStorageException, IllegalArgumentException {
+        if (chunks.containsKey(chunkName)) {
+            return new ChunkHandle(chunkName, false);
+        }
+        throw new ChunkNotFoundException(chunkName, "InMemoryChunkStorage::doOpenWrite");
+    }
+
+    @Override
+    protected int doRead(ChunkHandle handle, long fromOffset, int length, byte[] buffer, int bufferOffset) throws ChunkStorageException, NullPointerException, IndexOutOfBoundsException {
+        InMemoryChunk chunk = getInMemoryChunk(handle);
+        if (fromOffset >= chunk.getLength()) {
+            throw new ChunkStorageException(handle.getChunkName(), "Attempt to read at wrong offset");
+        }
+        if (length == 0) {
+            throw new ChunkStorageException(handle.getChunkName(), "Attempt to read 0 bytes");
+        }
+
+        // This is implemented this way to simulate possibility of partial read.
+        InMemoryChunkData matchingData = null;
+
+        // TODO : This is OK for now. This is just test code, but need binary search here.
+        for (InMemoryChunkData data : chunk.inMemoryChunkDataList) {
+            if (data.start <= fromOffset && data.start + data.length > fromOffset) {
+                matchingData = data;
+                break;
+
+            }
+        }
+
+        if (null != matchingData) {
+            int startIndex = (int) (fromOffset - matchingData.start);
+            byte[] source = matchingData.getData();
+            int i;
+            for (i = 0; i < length && bufferOffset + i < buffer.length && startIndex + i < source.length; i++) {
+                buffer[bufferOffset + i] = source[startIndex + i];
+            }
+            return i;
+        }
+        throw new ChunkStorageException(handle.getChunkName(), "No data was read");
+    }
+
+    @Override
+    protected int doWrite(ChunkHandle handle, long offset, int length, InputStream data) throws IndexOutOfBoundsException, ChunkStorageException {
+        InMemoryChunk chunk = getInMemoryChunk(handle);
+        long oldLength = chunk.getLength();
+        if (offset != chunk.getLength()) {
+            throw new IndexOutOfBoundsException("Attempt to write at wrong offset");
+        }
+        if (length == 0) {
+            throw new IndexOutOfBoundsException("Attempt to write 0 bytes");
+        }
+        if (chunk.isReadOnly) {
+            throw new ChunkStorageException(handle.getChunkName(), "chunk is readonly");
+        }
+
+        ByteArrayOutputStream out = new ByteArrayOutputStream(length);
```

Comment:
maybe you could move this block into  `chunk.append`

Don't we have IO utility methods that read a given amount of bytes from an InputStream ?

Reply:
Chunk doesn't need to know about InputStreams.
I don't think we have such methods in this code. We mostly read and write from ByteBuffer or BufferView/ArrayView (defined in io.pravega.common.util)